pyomo/contrib/doe/measurements.py

```diff
@@ -25,252 +25,265 @@
+        values: a ``list`` containing values which has the same shape of flattened variables 
+            default choice is None, means there is no give nvalues 
+
+        Return
+        ------
+        if not defining values, return a set of variable names 
+        if defining values, return a dictionary of variable names and its value 
+        """
+
+        self._names(var_name, extra_index,time_index)
+
+        self._add_elements(var_name, extra_index=extra_index, time_index=time_index)
+        if values:
+            # this dictionary keys are special set, values are its value
+            #self.special_set_value = {}
+            #for i in range(len(self.special_set)):
+            #    self.special_set_value[self.special_set[i]] = values[i]
+            self.special_set_value = self._generate_dict(values)
+
+        return self.special_set
+    
+    def update_values(self, values):
         """
-        Generate a dictionary, keys are the variable names, values are the indexes of this variable.
-        For e.g., name_and_index = {'C': ['CA', 'CB', 'CC']}
+        Used for updating values
+
         Parameters
         ---------
-        all_info: a dictionary, keys are measurement variable names,
-                values are a dictionary, keys are its extra index, values are its measuring time points
-                values are a list of measuring time point if there is no extra index for this measurement
-            Note: all_info can be the self.measurement_all_info, but does not have to be it.
+         values: a ``list`` containing values which has the same shape of flattened variables 
+            default choice is None, means there is no give nvalues 
         """
-        measurement_name = list(all_info.keys())
-        # a list of measurement extra indexes
-        measurement_extra_index = []
-        # check if the measurement has extra indexes
-        for i in measurement_name:
-            if type(all_info[i]) is dict:
-                index_list = list(all_info[i].keys())
-                measurement_extra_index.append(index_list)
-            elif type(all_info[i]) is list:
-                measurement_extra_index.append(None)
-        # a dictionary, keys are measurement names, values are a list of extra indexes
-        self.name_and_index = dict(zip(measurement_name, measurement_extra_index))
-
-    def _generate_flatten_name(self, measure_name_and_index):
+        self.special_set_value = self._generate_dict(values)
+
+    def add_bounds(self, upper_bound=None, lower_bound=None):
         """
-        Generate measurement flattened names
+        add bounds
+
         Parameters
         ----------
-        measure_name_and_index: a dictionary, keys are measurement names, values are lists of extra indexes
+        upper_bound: a ``list`` of upper bounds 
+        lower_bound: a ``list `` of lower bounds
+        """
+        if upper_bound:
```

Comment:
It is moved to add_elements. `Upper_bounds` and `lower_bounds` now allows 3 kinds of inputs: if a list, they should be either a scalar number (will be broadcast to be the same length as names), a list (same length as names). I am confused what kind of dictionary you mentioned: the keys should be the string names, but user wouldn't know the string names at that time because this function is generating these names. And we need to check if the keys are the same as the generated names. Wouldn't a list be enough?